### PR TITLE
Add Factory Reset button and switch

### PR DESCRIPTION
**Pull request by anatoly-savchenkov**
# What does this implement/fix?

This PR adds Reset button and switch components (and underlying changes to Preferences code) which perform a restart with resetting all preferences to default values. This can be helpful in case of:
- problems with complex designs especially when many global variables are used to keep some state
- transferring of ownership/installation point of ESP based equipment
- resetting to factory conditions for devices natively coming with ESPHome-based firmware

ESP32 code still to be tested as I don't have access to any ESP32 system at this moment. Docs will be added once maintainers are conceptually ok to accept this change.

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Other

**Related issue or feature (if applicable):** fixes <link to issue>

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** esphome/esphome-docs#2280

## Test Environment

- [ ] ESP32
- [ ] ESP32 IDF
- [x] ESP8266

## Example entry for `config.yaml`:

```yaml
# Example config.yaml
switch:
  - platform: factory_reset
    name: "Living Room Restart (Default Settings)"

button:
  - platform: factory_reset
    name: "Living Room Restart (Default Settings)"
```

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [ ] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

### CODE REVIEW

**Review by jesserockz — COMMENTED** [review]
I agree with the concept now that we have pre-flashed devices.

I wouldn't add a `switch` though, just keep the `button` (switches for shutdown/restart etc came before `button` was a thing so they have hung around)

I will give it a test and report back.

**Comment by anatoly-savchenkov** [comment]
My thinking was the following: buttons are still rare, switches are very common. New device class adds ~1KB of code. So, adding buttons support just to add this Reset component might be too heavy for devices with 512KB of flash.

Do you have plans to retire Shutdown/Restart switches?

**Comment by jesserockz** [comment]
Ah, I did not think about that. I have the restart button on every device via my common configs =)
Happy to keep both then.

File: esphome/components/esp8266/preferences.cpp [hunk 1]
```diff
@@ -243,17 +243,36 @@ class ESP8266Preferences : public ESPPreferences {
...
+    {
+      InterruptLock lock;
+      erase_res = spi_flash_erase_sector(get_esp8266_flash_sector());
+    }
+    if (erase_res != SPI_FLASH_RESULT_OK) {
+      ESP_LOGE(TAG, "Erase ESP8266 flash failed!");
+      return false;
+    }
+    /* memset(this->s_flash_storage, 0, sizeof(this->s_flash_storage[0]) * ESP8266_FLASH_STORAGE_SIZE);
+    s_flash_dirty = false; */
```
Comment (jesserockz) [inline esphome/components/esp8266/preferences.cpp]: What is this for?

**Review by jesserockz — CHANGES_REQUESTED** [review]
Generally looks good. One question above. 

Can we rename the platform to `factory_reset` though please? 
the terms reset/reboot/restart can be easily mistaken.

File: esphome/components/esp8266/preferences.cpp (hunk 1 above)
Comment (anatoly-savchenkov) [reply esphome/components/esp8266/preferences.cpp]: This is old code, should be removed

**Review by jesserockz — APPROVED** [review]

**Comment by jesserockz** [comment]
Please add a documentation PR, then this can be merged

File: esphome/components/factory_reset/switch/__init__.py [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+    "FactoryResetSwitch", switch.Switch, cg.Component
+)
+
+CONFIG_SCHEMA = switch.SWITCH_SCHEMA.extend(
+    {
+        cv.GenerateID(): cv.declare_id(FactoryResetSwitch),
+        cv.Optional(CONF_INVERTED): cv.invalid(
+            "Factory Reset switches do not support inverted mode!"
+        ),
+        cv.Optional(CONF_ICON, default=ICON_RESTART_ALERT): switch.icon,
```
Comment (jesserockz) [inline esphome/components/factory_reset/switch/__init__.py]:
```suggestion
        cv.Optional(CONF_ICON, default=ICON_RESTART_ALERT): cv.icon,
```

**Review by jesserockz — APPROVED** [review]

**Pull request merged**